pythonforandroid/pythonpackage.py

```diff
@@ -0,0 +1,664 @@
+""" This module offers highlevel functions to get package metadata
+    like the METADATA file, the name, or a list of dependencies.
+
+    Usage examples:
+
+       # Getting package name from pip reference:
+       from pytonforandroid.pythonpackage import get_package_name
+       print(get_package_name("pillow"))
+       # Outputs: "Pillow" (note the spelling!)
+
+       # Getting package dependencies:
+       from pytonforandroid.pythonpackage import get_package_dependencies
+       print(get_package_dependencies("pep517"))
+       # Outputs: "['pytoml']"
+
+       # Get package name from arbitrary package source:
+       from pytonforandroid.pythonpackage import get_package_name
+       print(get_package_name("/some/local/project/folder/"))
+       # Outputs package name
+
+    NOTE:
+
+    Yes, this module doesn't fit well into python-for-android, but this
+    functionality isn't available ANYWHERE ELSE, and upstream (pip, ...)
+    currently has no interest in taking this over, so it has no other place
+    to go.
+    (Unless someone reading this puts it into yet another packaging lib)
+
+    Reference discussion/upstream inclusion attempt:
+
+    https://github.com/pypa/packaging-problems/issues/247
+
+"""
+
+
+from io import open  # needed for python 2
+import os
+from pep517.envbuild import BuildEnvironment
+from pep517.wrappers import Pep517HookCaller
+import pytoml
+import shutil
+import subprocess
+import sys
+import tarfile
+import tempfile
+import textwrap
+import time
+try:
+    from urllib.parse import urlparse
+except ImportError:  # Python 2...
+    from urlparse import urlparse
+import zipfile
+
+
+def transform_dep_for_pip(dependency):
+    if dependency.find("@") > 0:
+        # For some reason, pip hates pep 508 urls.
+        # Translate to an #egg-name= reference:
+        url = (dependency.partition("@")[2].strip() +
+               "#egg-name=" +
+               dependency.partition("@")[0].strip()
+              )
+        return url
+    return dependency
+
+
+def extract_metainfo_files_from_package(
+        package,
+        output_folder,
+        debug=False
+        ):
+    """ Extracts metdata files from the given package to the given folder,
+        which may be referenced in any way that is permitted in
+        a requirements.txt file or install_requires=[] listing.
+
+        Current supported metadata files that will be extracted:
+
+        - pytoml.yml  (only if package wasn't obtained as wheel)
+        - METAINFO
+    """
+
+    if package is None:
+        raise ValueError("package cannot be None")
```

Comment:
I'm curious why don't you let it crash later with `AttributeError` rather than doing this?

Reply:
meh I usually don't do this but this function is a bit bigger and already does stuff in the filesystem before that would first cause an error, and I once passed None from another function and was confused :smile: no *good* reasons but it feels more appropriate to catch it early for this one :woman_shrugging: (I'm happy to remove it if you think otherwise)

Reply:
Yes I think that's a very fair point :+1:  maybe just add a comment stating so.
```
# sanity check to fail fast rather than in the middle of the process of touching the file system
```
With a unit test that prove the point it would obviously be even better. The test would not simply show that `ValueError` is being raised. But also that some critical part of the code (e.g. the ones touching the file system) are not being reached. You can use unittest.mock to achieve this.

And as a general note, anything that doesn't look natural in the code should have at least a comment if not a unit test